Filter an array containing the file to be updated

MarkGibbons opened the pull request:
Work in progress to show how I think using a ruby proc as a filter to do arbitrary things with a file might work.  Related to #95 and #79.  Took some of the ideas from poise-line.

Include filters for inserting lines before or after a match.
Introduces a new resource filter_lines. 

Attributes:
property :filter, [Method, Proc] - a proc or a method.  I've put some default methods in Line::Filter
property :filter_args, Array - arguments specific to the filter, might be lines to insert, regex, anything really
property :path, String - file to change

For example the :before filter can be used to insert a bunch of lines immediately before a matching regex.  Arguments are pattern, array of lines to insert, match and insert (each, first only, last)

The tests show some inline procs create on the fly.

I don't really like sending the arguments as positional parms in an array.  I'm not sure about using a method instead of a proc.  It works but is there something in ruby that makes it a bad idea.

The :before and :after filters need some love. They are first working draft code and are a bit too much like fortran instead of ruby.  People could write procs and extend the Line::Filter class or use their own filter class.

Thanks,
  Mark

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sous-chefs/line-cookbook/101)
<!-- Reviewable:end -->

Commit:
Working example of append_if_no_line on windows

Commit:
Delete lines windows support

windows support added to

resources/add_to_list.rb
resources/append_if_no_line.rb
resources/delete_lines.rb
resources/replace_or_add.rb
resources/delete_from_list.rb

Commit:
delete-from-list tested

Commit:
Add the sensitive and backup properties for the internal file resources

Commit:
Handle empty file and last line edge cases for add_to_list

Commit:
Handle last line conditions and edge cases

Add a test to be sure empty files are left as empty files when they should be
Retest append_if_no_line, delete_from_list, delete_line, replace_or_add

Commit:
Add the ignore_missing option to append_if_no_line

Readme word-smithing.
Note the breaking changes in the changelog.
Code clean up for style consistency across providers

damacus committed:
Add AppVeyor configuration
Get chef --version working
Use kitchen directly
Remove chef 12 test remnants
Clean up test resource handler

damacus committed:
Add CircleCI Support (Danger)

Commit:
Version 2.0.0

Commit:
WIP Filter an array containing the file to be updated
Include filters for inserting lines before or after a match.
Add filter_line inspec tests

Commit:
WIP Filter an array containing the file to be updated
Include filters for inserting lines before or after a match.
Add filter_line inspec tests

Commit:
Consistent processing when the target file is missing.

Add the ignore_missing option
Set the default for ignore_missing to true
Add tests of processing missing files with and without ignore_missing set to false
This change restores the processing of files to the 1.0.0 behavior.  We flipped with with 1.1.0 release and have gone
back to the old behavior with 2.0.0

Commit:
Consistent handling of missing files

Commit:
Fix the missing file test.

Move the chefspec test recipes to spec/fixtures
Merge the missing file tests with the other tests for each provider

Commit:
Remove the test suites from .kitchen.dokken.yml

Test kitchen will merge the test suites from .kitchen.yml for the travis tests.

MarkGibbons commented on test/fixtures/cookbooks/test/recipes/filter_lines.rb:
```diff
@@ -0,0 +1,125 @@
+
+template '/tmp/dangerfile' do
+end
+
```
Some examples of what's possible.  Inspec tests to check the results are there too.

MarkGibbons commented on test/fixtures/cookbooks/test/recipes/filter_lines.rb:
```diff
@@ -0,0 +1,125 @@
+  sensitive true
+end
+
+filter_lines 'Insert lines before match' do
```
Insert one time example.  Insert before the first match.

MarkGibbons commented on test/fixtures/cookbooks/test/recipes/filter_lines.rb:
```diff
@@ -0,0 +1,125 @@
+  filter_args [match_pattern, insert_lines]
+end
+
+filter_lines 'Insert lines before match 2nd try' do
```
Idempotence check

MarkGibbons commented on test/fixtures/cookbooks/test/recipes/filter_lines.rb:
```diff
@@ -0,0 +1,125 @@
+  sensitive true
+end
+
+filter_lines 'Insert lines before match' do
```
Insert multiple times.  Before each match of the regex.

MarkGibbons commented on test/fixtures/cookbooks/test/recipes/filter_lines.rb:
```diff
@@ -0,0 +1,125 @@
+  source 'dangerfile.erb'
+end
+
+filter_lines 'Reverse line text' do
```
Arbitrary in line proc.  Can do anything with the array as long as a new array is returned.

MarkGibbons committed:
Missing file (#100)

* Move common methods to the Line::Helper module
* Raise error if file missing
* Read the contents of the target file
* Chef target file existence
* Remove the double negative when checking for file existence
* Create a method to set the default for new_resource.sensitive
* Change the resource boiler plate lines to use the same order

jasonwbarnett committed:
Add usage details for append_if_no_line resource (#104)

MarkGibbons committed:
Missing file (#103)

* Move common methods to the Line::Helper module

raise error if file missing
read the contents of the target file
chef target file existence

* Remove the double negative when checking for file existence

* Create a method to set the default for new_resource.sensitive

* Change the resource boiler plate lines to use the same order

* Cookstyle fixes

* Update change log

* Remove the OS helper module. Use platform_family? instead.

* Minor fix and redrive the build tests

jasonwbarnett committed:
Fix markdown rendering for trailing whitespace (#105)

damacus committed:
Fix markdown!

damacus committed:
Drop Chef 12 support. Prepare 2.0 for release

damacus committed:
Add CODEOWNERS and github templates

Commit:
Add CONTRIBUTING.md and TESTING.md

Add the meta documentation file to chefignore

Commit:
Add development notes

Commit:
Add development notes

damacus commented on libraries/filter.rb:
```diff
@@ -2,11 +2,13 @@
   class Filter
     def before(current, args)
-      # Insert a set of lines before each match of the pattern
+      # Insert a set of lines immediately before each match of the pattern - seems like this has limited usefulness.  Comments and blank like would screw it up.
```
there's a wee typo here

MarkGibbons replied:
The before and after filters as they stand are unlikely to make the final cut.  I'm trying to figure out something more useful.

Commit:
Add the filter_lines resource

The filter resource allows a proc to be supplied to edit the lines of a file
Add the before filter, used to insert lines before the first, last or every match of a pattern line.
Add the after filter, used to insert lines after the first, last or every match of a pattern line.
Add the filters option to use multiple filters for one filter_lines resource.
Add the replace filter

Commit:
Add the filter_lines resource

The filter resource allows a proc to be supplied to edit the lines of a file
Add the before filter, used to insert lines before the first, last or every match of a pattern line.
Add the after filter, used to insert lines after the first, last or every match of a pattern line.
Add the filters option to use multiple filters for one filter_lines resource.
Add the replace filter

damacus committed:
Add github lock bot configuration

Commit:
Add stanza processing

Limited stanza syntax
Need refactoring and a DRY review

Commit:
Add stanza processing

Limited stanza syntax
Need refactoring and a DRY review

Commit:
Move each filter to its own file

Commit:
Move each filter to its own file

Commit:
Add the between filter

Commit:
Add the between filter

Commit:
Fix the circleci testing notes

Commit:
Fix the circleci testing notes

Commit:
Add the comment and substitute filters

Commit:
Add the comment and substitute filters

Commit:
Handle error conditions

Commit:
Handle error conditions

Commit:
Add license headers

Commit:
Add license headers

Commit:
Verify the arguments passed to the filters

Commit:
Verify the arguments passed to the filters

Commit:
Add tests for the filter helper methods

Commit:
Add tests for the filter helper methods

Commit:
Version 2.1.0

Commit:
Version 2.1.0

MarkGibbons committed:
Sensitive: Test on chef 12. Minor fix for the sensitive attribute (#111)

* Sensitive: Test on chef 12. Minor fix for the sensitive attribute

* Undo the .travis.yml modifications

Commit:
Style fixes

Commit:
Style fixes

Commit:
Merge branch 'master' of https://github.com/sous-chefs/line-cookbook into proc

Commit:
Add the filter-lines test instances to travis

Commit:
Add the filter-lines test instances to travis

Commit:
Fix the before and after insert location filter. nil is valid.

Commit:
Fix the before and after insert location filter. nil is valid.

Commit:
Allow the replacement lines to be a single string

Commit:
Allow the replacement lines to be a single string

Commit:
Style fixes

Commit:
Style fixes

Commit:
Fix the README.md formatting

Commit:
Fix the README.md formatting

Commit:
Fix README formatting

Commit:
Fix README formatting

Commit:
Readme updates

Commit:
Readme updates

Commit:
Enable the resource execute conditional tests they were always allowing the resource to run

Commit:
Enable the resource execute conditional tests they were always allowing the resource to run

Commit:
Include links to other file editting cookbooks

Commit:
Include links to other file editting cookbooks

Commit:
Simplify the way filters are specified. Remove the filter and filter_args properties

Commit:
Simplify the way filters are specified. Remove the filter and filter_args properties

Commit:
Clean up the readme

Commit:
Clean up the readme

Commit:
Fix README

Commit:
Fix README

Commit:
Fix spelling

Commit:
Fix spelling

Commit:
Include an example of using each filter in the readme.

Commit:
Include an example of using each filter in the readme.

Commit:
Style fix

Commit:
Style fix

Commit:
Add the delete_between filter

Commit:
Add the delete_between filter

Commit:
Add delete_between to the Changelog for Version 2.1.0

Commit:
Add delete_between to the Changelog for Version 2.1.0

Commit:
Merge branch 'proc' of https://github.com/MarkGibbons/line-cookbook into proc

Commit:
Merge branch 'master' of https://github.com/sous-chefs/line-cookbook into proc
Add the 2.0.2 branch

Commit:
Fix merge error

damacus reviewed (commented):
> Reviewed 3 of 31 files at r7, 1 of 8 files at r10, 1 of 5 files at r12.
> *[Reviewable](https://reviewable.io/reviews/sous-chefs/line-cookbook/101)* status: 5 of 47 files reviewed, all discussions resolved (waiting on @MarkGibbons)



<!-- Sent from Reviewable.io -->
